src/TraceEvent/Symbols/AdalAuth.cs

```diff
@@ -0,0 +1,59 @@
+﻿#if NET45
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Microsoft.IdentityModel.Clients.ActiveDirectory;
+
+namespace Microsoft.Diagnostics.Symbols
+{
+    public static class Adal
+    {
+        private static object s_lock = new object();
+        private static DateTime s_lastAttempt = DateTime.MinValue;
+        private static string s_accessToken;
+
+        public static string AcquireToken()
+        {
+            lock (s_lock)
+            {
+                if (s_accessToken != null)
+                {
+                    return s_accessToken;
+                }
+
+                // Do not prompt more than once/minute
+                var now = DateTime.UtcNow;
+                if (now < s_lastAttempt + TimeSpan.FromSeconds(60))
+                {
+                    return null;
+                }
+                s_lastAttempt = now;
+            }
+
+            var tenantId = "microsoft.com";
+            var authority = $"https://login.microsoftonline.com/{tenantId}";
+            var resourceId = "499b84ac-1321-427f-aa17-267ca6975798"; // ADO
+
+            var clientId = "1950a258-227b-4e31-a9cf-717495945fc2"; // Azure PowerShell
+            var redirectUri = new Uri("urn:ietf:wg:oauth:2.0:oob");
+
+            var authContext = new Microsoft.IdentityModel.Clients.ActiveDirectory.AuthenticationContext(authority);
+
+            try
+            {
+                var token = authContext.AcquireTokenAsync(resourceId, clientId, redirectUri, new PlatformParameters(PromptBehavior.Auto)).Result;
+                lock (s_lock)
+                {
+                    s_accessToken = token.AccessToken;
```

Comment:
You also need to record the expiration time so that you know when to refresh the token.

Reply:
Done